### PR TITLE
Use ASWebAuthenticationSession on iOS 12

### LINKED ISSUE #238: SFAuthenticationSession is now deprecated in iOS 12, ASWebAuthenticationSession should be used instead

### PR DESCRIPTION
**Pull request by martinoconnor**
Fixes #238

### CODE REVIEW

**Comment by cocojoe** [comment]
Thanks for the PR @martinoconnor

File: Auth0/SafariWebAuth.swift [hunk 1]
```diff
@@ -120,7 +120,11 @@ class SafariWebAuth: WebAuth {
         let authorizeURL = self.buildAuthorizeURL(withRedirectURL: redirectURL, defaults: handler.defaults, state: state)
 
         #if swift(>=3.2)
-        if #available(iOS 11.0, *), self.authenticationSession {
+        if #available(iOS 12.0, *), self.authenticationSession {
+            let session = ASAuthenticationSession(authorizeURL: authorizeURL, redirectURL: redirectURL, state: state, handler: handler, finish: callback, logger: logger)
+            logger?.trace(url: authorizeURL, source: "AuthenticationServicesAuthenticationSession")
+            self.storage.store(session)
+        } else if #available(iOS 11.0, *), self.authenticationSession {
```
Comment (cocojoe) [inline Auth0/SafariWebAuth.swift]: For iOS 11 you still have to use SFAuthenticationSession as per Apple guidelines. So there are 3 tiers of support iOS >=12, iOS 11, iOS <11.0

File: Auth0/ASAuthenticationSession.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import UIKit
+import AuthenticationServices
+
+#if swift(>=3.2)
+@available(iOS 12.0, *)
+class ASAuthenticationSession: AuthSession {
```
Comment (cocojoe) [inline Auth0/ASAuthenticationSession.swift]: Do we really need a whole new class or could we `typealias` the correct method based upon iOS?

Comment (martinoconnor) [reply Auth0/ASAuthenticationSession.swift]: I think the class is probably justified as the ASWebAuthenticationSession or SFAuthenticationSession is stored in a member variable of the respective class. Then we have to worry about the error type in the handler callback which is related to the corresponding session type

**Comment by martinoconnor** [comment]
Looks like the build failure is due to CocoaPods/CocoaPods#7708 so will have to wait until 1.6 of CocoaPods comes out of beta

File: Auth0/SafariWebAuth.swift (hunk 1 above)
Comment (cocojoe) [reply Auth0/SafariWebAuth.swift]: My mistake, saw the word `Safari`, naming these classes/methods is quite tricky 😆

**Comment by cocojoe** [comment]
#240 This will fix the pod lint issue.

Please can you rebase this PR against master. Thx

File: Auth0/ASAuthenticationSession.swift (hunk 2 above)
Comment (martinoconnor) [reply Auth0/ASAuthenticationSession.swift]: I have moved the changes inside the original SafariAuthenticationSession class

**Comment by cocojoe** [comment]
@martinoconnor what about https://github.com/auth0/Auth0.swift/blob/master/Auth0/WebAuth.swift#L244 ?

File: Auth0/SafariAuthenticationCallback.swift [hunk 3]
```diff
@@ -22,22 +22,28 @@
...
 import UIKit
 import SafariServices
+import AuthenticationServices
 
 #if swift(>=3.2)
 @available(iOS 11.0, *)
 class SafariAuthenticationSessionCallback: AuthTransaction {
 
     var state: String?
-    var authSession: SFAuthenticationSession?
```
Comment (cocojoe) [inline Auth0/SafariAuthenticationCallback.swift]: By removing this, there is nothing retaining the session alive, so as soon as you accept the `Auth0.com` dialog it will fail and never launch the browser.

`Attempting to load the view of a view controller while it is deallocating is not allowed and may result in undefined behavior (<SFAuthenticationViewController: 0x7f851c82b600>)`

File: Auth0/SafariAuthenticationCallback.swift [hunk 4]
```diff
@@ -22,22 +22,28 @@
...
     var callback: (Bool) -> Void = { _ in }
 
     init(url: URL, schemeURL: String, callback: @escaping (Bool) -> Void) {
         self.callback = callback
-        self.authSession = SFAuthenticationSession(url: url, callbackURLScheme: schemeURL) { [unowned self] url, _ in
```
Comment (cocojoe) [inline Auth0/SafariAuthenticationCallback.swift]: Related to previous comment.

File: Auth0/SafariAuthenticationSession.swift [hunk 5]
```diff
@@ -22,31 +22,39 @@
...
 
-    typealias FinishSession = (Result<Credentials>) -> Void
-
-    var authSession: SFAuthenticationSession?
-    let authorizeURL: URL
-
     init(authorizeURL: URL, redirectURL: URL, state: String? = nil, handler: OAuth2Grant, finish: @escaping FinishSession, logger: Logger?) {
-        self.authorizeURL = authorizeURL
         super.init(redirectURL: redirectURL, state: state, handler: handler, finish: finish, logger: logger)
-        self.authSession = SFAuthenticationSession(url: self.authorizeURL, callbackURLScheme: self.redirectURL.absoluteString) { [unowned self] in
```
Comment (cocojoe) [inline Auth0/SafariAuthenticationSession.swift]: Retention doesn't appear to be an issue here, although I'm sure it was hence the use of the `authSession` property.
It's difficult to test these behaviour in unit tests.

Comment (martinoconnor) [reply Auth0/SafariAuthenticationSession.swift]: Looks like SFAuthenticationSession requires a strong reference, as per the apple docs

> Note
> 
> Ensure that there is a strong reference to the SFAuthenticationSession instance when the session is in progress.

I don't see a similar note for ASWebAuthenticationSession

Comment (martinoconnor) [reply Auth0/SafariAuthenticationSession.swift]: If ASWebAuthenticationSession is deallocated, the permissions dialog is dismissed

File: Auth0/SafariAuthenticationCallback.swift (hunk 3 above)
Comment (cocojoe) [reply Auth0/SafariAuthenticationCallback.swift]: AnyObject not work here? Doesn't look right pulling in ObjC

File: .circleci/config.yml [hunk 6]
```diff
@@ -40,4 +40,4 @@ workflows:
   version: 2
   build-test-report:
     jobs:
-      - build-and-test
```
Comment (cocojoe) [inline .circleci/config.yml]: It would be good not to make unnecessary formatting changes thanks

File: Auth0/SafariAuthenticationCallback.swift (hunk 4 above)
Comment (cocojoe) [reply Auth0/SafariAuthenticationCallback.swift]: Same as previous `NSObject` comment

File: fastlane/Scanfile [hunk 7]
```diff
@@ -1,5 +1,5 @@
 scheme "Auth0.iOS"
-devices ["iPhone 8"]
+devices ["iPhone 8 (11.2)", "iPhone 8 (12.0)", "iPhone 7 (10.3.1)"]
```
Comment (cocojoe) [inline fastlane/Scanfile]: I see what you're doing for your own tests, but please don't include in this PR. At best it would be in a different PR, although I intend to replace with multiple platform tests, which i recently added to Lock.swift.

File: Auth0/SafariAuthenticationSession.swift [hunk 8]
```diff
@@ -22,31 +22,45 @@
...
-            guard $1 == nil, let callbackURL = $0 else {
-                if case SFAuthenticationError.canceledLogin = $1! {
-                    self.finish(.failure(error: WebAuthError.userCancelled))
-                } else {
-                    self.finish(.failure(error: $1!))
+        if #available(iOS 12.0, *) {
+            let authSession = ASWebAuthenticationSession(url: authorizeURL, callbackURLScheme: self.redirectURL.absoluteString) { [unowned self] in
```
Comment (cocojoe) [inline Auth0/SafariAuthenticationSession.swift]: Seems like a lot of duplication of code in side these blocks, could we refactor this down to one block and a typealias?

**Comment by cocojoe** [comment]
Also please rebase again master, thanks.

**Comment by martinoconnor** [comment]
Apologies for the delay, work has been busy lately, but will make the changes when I'm free

**Comment by cocojoe** [comment]
@martinoconnor no problem, let me know if you need a hand on finishing this up.

**Comment by cocojoe** [comment]
Moved to #245
